torch/lib/THD/base/init_methods/InitMethodFile.cpp

```diff
@@ -87,88 +80,135 @@ InitMethod::Config initFile(std::string file_path, rank_type world_size,
-  file << group_name << ' ' << assigned_rank << ' ' << port
-       << ' ' << ifa_addresses.size();
-  for (auto addr_str : ifa_addresses) {
-    file << ' ' << addr_str;
-  }
-  file << std::endl;
+  return {fd, std::count(content.begin(), content.end(), '\n')};
+}
 
-  std::size_t lines = 0; // we have just added new line
-  while (lines < world_size) { // wait until all processes will write their info
+std::size_t waitForData(int fd, std::fstream& file, rank_type world_size) {
+  std::size_t lines = 0;
+  // Wait until all processes will write their info
+  while (lines < world_size) {
     unlockFile(fd);
     std::this_thread::sleep_for(std::chrono::milliseconds(200));
     lockFile(fd);
 
     file.sync();
     file.seekp(0, std::ios_base::beg);
-    content = {std::istreambuf_iterator<char>(file),
-               std::istreambuf_iterator<char>()};
+    std::string content = {std::istreambuf_iterator<char>(file),
+                           std::istreambuf_iterator<char>()};
     lines = std::count(content.begin(), content.end(), '\n');
-    file.seekp(0, std::ios_base::beg);
   }
-  // NOTE: the loop exits with a locked fd
 
+  file.seekp(0, std::ios_base::beg);
+  return lines;
+}
+
+// master_port, master_addrs, ranks
+std::tuple<port_type, std::vector<std::string>, std::vector<int>>
+parseFile(std::fstream& file, rank_type world_size, std::string group_name) {
   port_type master_port;
-  std::vector<std::string> master_addresses;
+  std::vector<std::string> master_addrs;
   std::vector<int> ranks(world_size);
+  // Parse the file
   for (std::size_t i = 0; i < world_size; ++i) {
-    std::string file_group_name, tmp_address;
-    std::size_t addresses_count;
-    int rank;
-    port_type tmp_port;
-
-    file >> file_group_name >> rank >> tmp_port >> addresses_count;
-    if (file_group_name != group_name) {
-      throw std::logic_error("file_group_name != group_name");
+    std::string proc_group_name;
+    std::size_t proc_addrs_count;
+    int proc_rank;
+    port_type proc_port;
+
+    file >> proc_group_name >> proc_rank >> proc_port >> proc_addrs_count;
+    if (proc_group_name != group_name) {
+      throw std::logic_error("proc_group_name != group_name");
     }
 
-    std::vector<std::string> tmp_addresses(addresses_count);
-    for (std::size_t j = 0; j < addresses_count; ++j) {
-      file >> tmp_address;
-      tmp_addresses.emplace(tmp_addresses.begin() + j, tmp_address);
+    std::vector<std::string> proc_addrs(proc_addrs_count);
+    for (auto& str : proc_addrs) {
+      file >> str;
     }
 
-    ranks[i] = rank;
+    ranks[i] = proc_rank;
     // Whether there is already assigned rank 0, or we have to get addresses and port
     // from first process which has unassigned rank (it will be rank 0).
-    if (rank == 0 || (rank < 0 && master_addresses.size() == 0)) {
-      master_port = tmp_port;
-      master_addresses = tmp_addresses;
+    if (proc_rank <= 0 && (proc_rank == 0 || master_addrs.size() == 0)) {
```

Comment:
Actually can you revert it to be `proc_rank == 0 || (proc_rank == -1 && master_addrs.size() == 0)`? It seems clearer to me now